### PR TITLE
Update ZIO library

### LINKED ISSUE #6069: Update ZIO library
## Problem

ZIO had a major update from version 1 to 2 that changed the API. The library is used by the language server and the project manager. 

The update should also help with the #6028 issue

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #6069 

Changelog:
- update: ZIO major version bump

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala [hunk 1]
```diff
@@ -18,7 +17,7 @@ trait Exec[-F[_, _]] {
     * @param op effect to execute
     * @return a future containing either a failure or a result
     */
-  def exec[E, A](op: F[E, A]): Future[Either[E, A]]
+  def exec[E, A](op: F[E, A])(implicit trace: Trace): Future[Either[E, A]]
```
Comment (hubertp) [inline engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:20]: We seem to have documentation for the other parameters, so probably `trace` would be good as well (and all the other methods defining that parameter)

**Review by hubertp — APPROVED** [review]

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala [hunk 2]
```diff
@@ -47,55 +46,66 @@ case class ZioExec(runtime: effect.Runtime) extends Exec[ZioExec.IO] {
...
-  override def exec[E, A](op: ZIO[ZEnv, E, A]): Future[Either[E, A]] = {
-    val promise = Promise[Either[E, A]]()
-    runtime.instance.unsafeRunAsync(op) {
-      _.fold(
-        ZioExec.completeFailure(promise, _),
-        ZioExec.completeSuccess(promise, _)
+  override def exec[E, A](
+    op: ZIO[ZAny, E, A]
+  )(implicit trace: Trace): Future[Either[E, A]] =
+    zio.Unsafe.unsafe { implicit unsafe =>
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:55]: [zio runtime](https://zio.dev/reference/core/runtime/) says:

> Whenever we run an unsaferun function

Why the heck is that called `unsafe`? What's unsafe on running something, expect it runs!?

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala [hunk 3]
```diff
@@ -47,55 +46,66 @@ case class ZioExec(runtime: effect.Runtime) extends Exec[ZioExec.IO] {
...
+      val promise = Promise[Either[E, A]]()
+      runtime.instance.unsafe.fork(
```
Comment (JaroslavTulach) [inline engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:57]: > have to spawn a fiber every time we call fork on an effect to spawn off a new fiber.

OK, so now we create new thread pool, right? Previously we just `unsafe`(!)`RunAsync` - e.g. sent the execution to an existing thread pool.

**Review by JaroslavTulach — APPROVED** [review]
I learned a bit about ZIO while reviewing this PR, however my approval is still more a "process approval" than code approval.

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala (hunk 3 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:57]: The logic didn't change, just the API. Fiber is like a green thread. All the computations are still happening in the thread pool configured on the startup.

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala (hunk 2 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:55]: There is a way to write your programs when you don't execute the effects right away but compose them into a bigger computation. In the zio library, the `ZIO` type represents such a program. The only safe way to run the `ZIO` computation is to return it from the "main" (in zio case it is called `run`) function:

https://github.com/enso-org/enso/blob/6b9cbeacb2caa7606d91fda90ec3ec40acb39c70/lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/ProjectManager.scala#L105

It was inspired by pure functional languages like Haskell, where the language simply doesn't have capabilities to perform effects other than returning them from the main function https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO.html#t:IO and have a similar backdoor https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO-Unsafe.html

Comment (JaroslavTulach) [reply engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:55]: I see. I know how actions (IO operations) are done in Haskell. I didn't know about `unsafePerformIO` in Haskell - I do understand why it is called **unsafe** in Haskell.

On the other hand there doesn't seem anything **unsafe** (from JVM perspective) on launching an effect and waiting for its result (if done outside of the effects worker pool). It is certainly not as **unsafe** as using `sun.misc.Unsafe`.

Thanks for the explanation and pointers.

File: engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala (hunk 3 above)
Comment (JaroslavTulach) [reply engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:57]: There are no _green threads_ in the JVM we are using. [Project Loom](https://openjdk.org/projects/loom/) has only become an experimental feature of JDK 19 - without it there is no chance to have any _green threads_ in JVM.

Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/effect/Exec.scala:57]: Correct. That's why ZIO has its own primitive called [`Fiber`](https://zio.dev/reference/fiber/).

**Review by radeusgd — APPROVED** [review]
Approving the changes to files owned by me - here it's the legal review report. The updated review looks OK.

**Pull request merged**